using Data.TransferObject.ViewModel;
using System;
using System.Linq;
using System.Security.Claims;
using System.Threading.Tasks;
using System.Web;
using System.Web.Http;
using Wrapper.Admin;


namespace careapi.Controllers
{
    [RoutePrefix("api/v1/requiredFee")]
    [Authorize]
    public class RequiredFeeController : ApiController
    {
        private readonly IWrapperAdminRepository _wrapper;
        public RequiredFeeController(IWrapperAdminRepository wrapper)
        {
            _wrapper = wrapper;
        }

        [HttpPost]
        [Route("")]
        public async Task<IHttpActionResult> Create([FromBody]RequiredFeeVM model)
        {
            try
            {
           
                var result = await _wrapper.RequiredFee.CreateRequiredFeeAsync(model);

                if (result != null)
                {
                     return Ok(new { response = result, success = true });
                }
            }
            catch (Exception ex)
            {
                return Ok( new { success = false, error = ex.Message, response = "Error has accoured trying to process your request" });
            }
             return Ok(new { response = "Record not found", success = false });
        }

        [HttpGet]
        [Route("")]
        public async  Task<IHttpActionResult> Get()
        {
            try
            {
              
                var result = await _wrapper.RequiredFee.GetRequiredFee();

                if (result != null)
                {
                    return Ok(new { response = result, success = true });
                }
            }
            catch (Exception ex)
            {
                return Ok( new { success = false, error = ex.Message, response = "Error has accoured trying to process your request" });
            }
             return Ok(new { response = "Record not found", success = false });
        }


		
        [HttpGet]
        [Route("{requiredFeeId}/requiredFeeId")]
        public async Task<IHttpActionResult> Get(int requiredFeeId)
        {
            try
            {   
                                   var tokenIdentity = new ClaimsIdentity(HttpContext.Current.User.Identity);
                var decryptedToken = tokenIdentity.Claims;

                var tenant = decryptedToken.FirstOrDefault(st => st.Type == "tenantId").Value;

                int tenantId = 0;

                if (tenant != null)
                    tenantId = Convert.ToInt32(tenant);

                var result = await _wrapper.RequiredFee.GetRequiredFee(requiredFeeId);

                if (result != null)
                {
                    return Ok(new { response = result, success = true });
                }
            }
            catch (Exception ex)
            {
                return Ok( new { success = false, error = ex.Message, response = "Error has accoured trying to process your request" });
            }
            return Ok(new { response = "Record not found", success = false });
        }

        [HttpPut]
		[Route("{requiredFeeId}/requiredFeeId")]
        public async Task<IHttpActionResult> Update([FromBody]RequiredFeeVM model, int requiredFeeId)
        {
            try
            {
                var result = await _wrapper.RequiredFee.UpdateRequiredFee(requiredFeeId, model);

                if (result != false)
                {
                    return Ok(new { response = result, success = true });
                }
            }
            catch (Exception ex)
            {
                return Ok( new { success = false, error = ex.Message, response = "Error has accoured trying to process your request" });
            }
             return Ok(new { response = "Record not found", success = false });
        }
        [HttpDelete]
		[Route("{requiredFeeId}/requiredFeeId")]
        public async Task<IHttpActionResult> Delete(int requiredFeeId)
        {
            try
            {
                var result = await _wrapper.RequiredFee.DeleteRequiredFee(requiredFeeId);

                if (result != false)
                {
                    return Ok(new { response = result, success = true });
                }
            }
            catch (Exception ex)
            {
                return Ok( new { success = false, error = ex.Message, response = "Error has accoured trying to process your request" });
            }
            return Ok(new { response = "Record not found", success = false });
        }
    }
}


     
